src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -16,40 +16,88 @@
 import io.supertokens.pluginInterface.oauth.OAuthStorage;
 
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 public class OAuth {
 
-    public static void getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
-                                           String redirectURI, String responseType, String scope, String state)
-            throws InvalidConfigException {
+    private static final String LOCATION_HEADER_NAME = "Location";
+    private static final String COOKIES_HEADER_NAME = "Set-Cookie";
+
+    public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier appIdentifier, Storage storage, String clientId,
+                                                        String redirectURI, String responseType, String scope, String state)
+            throws InvalidConfigException, HttpResponseException, IOException, OAuthException, StorageQueryException,
+            TenantOrAppNotFoundException {
         // TODO:
         // - validate that client_id is present for this tenant
-        // - call hydra
-        // - if location header is:
-        //     - localhost:3000, then we redirect to apiDomain
-        //     - public url for hydra, then we throw a 400 error with the right json
-        //     - else we redirect back to the client
 
         OAuthStorage oauthStorage = StorageUtils.getOAuthStorage(storage);
 
         String redirectTo = null;
+        List<String> cookies = null;
+
+        String publicOAuthProviderServiceUrl = Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl();
 
         if (!oauthStorage.doesClientIdExistForThisApp(appIdentifier, clientId)) {
-            redirectTo = Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl() +
+            redirectTo =  publicOAuthProviderServiceUrl +
                     "/oauth2/fallbacks/error?error=invalid_client&error_description=Client+authentication+failed+%28e" +
                     ".g.%2C+unknown+client%2C+no+client+authentication+included%2C+or+unsupported+authentication" +
                     "+method%29.+The+requested+OAuth+2.0+Client+does+not+exist.";
         } else {
             // we query hydra
+            Map<String, String> queryParamsForHydra = constructHydraRequestParams(clientId, redirectURI, responseType, scope, state);
+            Map<String, String> responseHeaders = new HashMap<>();
+            HttpRequest.sendGETRequestWithResponseHeaders(main, null, Config.getBaseConfig(main).getOAuthProviderPublicServiceUrl(), queryParamsForHydra, 20, 400, null, responseHeaders); // TODO is there some kind of config for the timeouts?
+
+            if(!responseHeaders.isEmpty() && responseHeaders.containsKey(LOCATION_HEADER_NAME)) {
+                String locationHeaderValue = responseHeaders.get(LOCATION_HEADER_NAME);
+
+                if (locationHeaderValue.equals(publicOAuthProviderServiceUrl)){
```

Comment:
We have to be careful here:
- There could be a trailing `/` configured by the user in the config for the url, and hydra may not return that in the locationHeaderValue.
- `locationHeaderValue` returned from hydra is going to have a path and query params and everything int he url, whereas `publicOAuthProviderServiceUrl` configured by the user will not have any path. So the `.equals` will never be true.
- If the user configures `publicOAuthProviderServiceUrl` as a service name (for example if they are using docker compose), then the url returned from hydra response won't match. Maybe we need a different core config like `publicOAuthProviderServiceUrlConfiguredInHydra` - which by default will be the same as the `publicOAuthProviderServiceUrl` config. And we can also assert in validateAndNormalise functino (on coreConfig.java) that `publicOAuthProviderServiceUrlConfiguredInHydra` always resolves to a URL.

Reply:
I see that you used `.contains` now instead of `.equals`, which is better. However, it could cause issues if the `locationHeaderValue` returned from hydra is something like `https://example.com?...` and the value configured for `publicOAuthProviderServiceUrl` is `https://example.com/`. The difference here is the trailing `/`. 

It may be a better idea to normalise URLS first before checking comparing them. One way could be to parse them through some lib that returns a URL object, and then fetch the domain from it and then compare that string with `.equals`. 

URL comparison is tricky, and we need to be careful.

-------

Also, you didn't reply to the docker server point I bought up above - what do you think?

Reply:
I have to dig deeper into hydra, to answer those. Yesterday I didn't get to it. Now I feel like I am missing some important details. I will do it today.

Reply:
So, regarding the docker service address: the url used by hydra here is the default fallback for now (to `{hydra_self}/oauth2/fallbacks/error`). This indeed can be different from what was set in the core config as publicOAuthProviderServiceUrl, although this could be resolved in the hydra settings too. On the other hand, I would suggest to handle this on the core side, as we have greater control over that.